#167962621 Build User Registration Endpoint

hameedhub opened the pull request:
**What does this PR do?**

 Have user register an account

  **Description of Task to be completed?**
 - Have user registration endpoint working

   <code>POST /api/v1/auth/signup </code>

 - Created signup controller
 - Installed bcryptjs and jsonwebtoken
 - Created encryptor service
 - Created tokenGenerator service
 - Have test database on ElephantSQL
 - Updated README.md file
 - Swagger documentation of the Endpoint
 - Wrote Test cases

  **How should this be manually tested?**
   - After cloning the repo, cd into it and RUN <code>npm test </code> on terminal
   - Using Postman test endpoint RUN <code>npm run server</code> then send the above endpoint

  **Any background context you want to provide?**

  None

  **What are the relevant stories?**

  #167962621

  **Screenshots (if appropriate)**
 
![Screen Shot 2019-08-22 at 11 23 56 PM](https://user-images.githubusercontent.com/46590803/63554071-50700280-c534-11e9-9b59-672acc25e6b1.png)

![Screen Shot 2019-08-22 at 11 25 44 PM](https://user-images.githubusercontent.com/46590803/63554135-801f0a80-c534-11e9-892c-57ab0fdbf170.png)

![Screen Shot 2019-08-22 at 11 25 02 PM](https://user-images.githubusercontent.com/46590803/63554156-962ccb00-c534-11e9-8eaf-abe72d8c1774.png)



  **Questions:**

 None

Commit:
feature: User Login
[finishes #167735036]

Commit:
running migration on travis

Commit:
running migration on travis

Commit:
running migration on travis

Commit:
running migration on travis

Commit:
running migration on travis

Commit:
running migration on travis

Commit:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
running migration on travis

SamuelKoroh committed:
fixing travis build error

SamuelKoroh committed:
User can login

hameedhub committed:
feature(documentation):update README.md and swagger doc
[Finishes #167962621]

SamuelKoroh committed:
ft: user can login
- finished implementating feedback

SamuelKoroh committed:
finsihed implementing the feedback

SamuelKoroh committed:
Update .eslintrc

SamuelKoroh committed:
Update .gitignore

hameedhub committed:
feature(authentication):implementation feedback

hameedhub committed:
feature(documentation):update README.md

SamuelKoroh commented on src/test/auth.spec.js:
```diff
@@ -0,0 +1,108 @@
+        .set('Content-Type', 'application/json')
+        .send(user2);
+      res.should.have.status(201);
+      res.body.data.should.have.property('token');
```
After each test case put a new line

SamuelKoroh commented on src/test/auth.spec.js:
```diff
@@ -0,0 +1,108 @@
+        .post(`${URL_PREFIX}/signup`)
+        .set('Content-Type', 'application/json')
+        .send(user2);
+      res.should.have.status(201);
```
After declaring a const give a new line

SamuelKoroh commented on src/services/authService.js:
```diff
@@ -0,0 +1,58 @@
+
+    const { dataValues: user } = result;
+    const payload = { id: user.id, is_admin: user.is_admin };
+    const token = Helper.genToken(payload);
```
Put a new line before the return statement

SamuelKoroh commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,61 @@
+   * @returns { string} - encypted data
+   */
+  static encryptor(data) {
+    const encrypted = bcrypt.hashSync(data, salt);
```
Put a newline before the return statement

SamuelKoroh commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,61 @@
+   * @returns {string} - token
+   */
+  static genToken(payloader) {
+    const token = jwt.sign(payloader, secret, { expiresIn: '1hr' });
```
Put a newline before the return statement

SamuelKoroh commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,61 @@
+   * @returns { string} - encypted data
+   */
+  static encryptor(data) {
+    const encrypted = bcrypt.hashSync(data, salt);
```
Why not use async like await bcrypt.hash(data, salt) before hashSync is blocks the thread

SamuelKoroh replied:
Here is an example to make it dynamic
So that this method job is solely for encryption instead of declaring the salt outside

static async encryptor(data, rounds=10) {
    const salt = await bcrypt.genSalt(rounds)
    const encrypted = await bcrypt.hash(data, salt);

    return encrypted;
  }

SamuelKoroh commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,61 @@
+import bcrypt from 'bcrypt';
+import jwt from 'jsonwebtoken';
+
+const salt = bcrypt.genSaltSync(10);
```
Move the salt to encrypting method and use const salt =await bcrypt.genSalt(10); instead

SamuelKoroh commented on src/services/authService.js:
```diff
@@ -0,0 +1,58 @@
+    const checkEmail = await User.findOne({ where: { email } });
+
+    if (checkEmail) throw new Error(`Email '${email}' already exists`);
+    const encryptedPassword = Helper.encryptor(password);
```
After modifying the encryptor you can call it by adding await

const encryptedPassword = await Helper.encryptor(password);

Still no need to pass rounds to it since rounds is 10 by default

Annmary12 commented on src/services/authService.js:
```diff
@@ -0,0 +1,58 @@
+   * @param {*} password - string
+   * @return {JSON} - json format
+   */
+  static async signup(first_name, last_name, email, password) {
```
@hameedhub , instead of passing the user info individually, put them in an object

Example: Instead of having this
`static async signup(first_name, last_name, email, password) {`

We can have this
`static async signup(userDetails) {`

`userDetails` object will contain
`{
first_name: '', 
last_name: '', 
email: '', 
password: ''
}`

hameedhub committed:
feedback implementation

Annmary12 commented on src/controllers/authController.js:
```diff
@@ -0,0 +1,39 @@
+      const data = await AuthService.signup({
+        first_name, last_name, email, password
+      });
+      res.status(201).json({ status: 'created', data });
```
Status should be `success` instead of `created`

Annmary12 commented on src/test/auth.spec.js:
```diff
@@ -0,0 +1,113 @@
+        .set('Content-Type', 'application/json')
+        .send({ email: 'email@email.com', password: 'password' });
+
+      res.should.have.status(400);
```
You can also assert that the status message is `error` or `success` depending on what it should return, please apply this in other test cases.

Annmary12 commented:
Please add your register documentation under auth session, you don't need to create yours because Samuel has created that already.

Also, fix hounds errors above.

hameedhub committed:
feedback implementation 2

Annmary12 commented on src/services/authService.js:
```diff
@@ -0,0 +1,66 @@
+   * @param {*} last_name - string
+   * @param {*} email - valid email
+   * @param {*} password - string
+   * @return {JSON} - json format
```
param is just userDetails which is an object not `lastname, email and password`

hameedhub committed:
implementation of feedback 3

SamuelKoroh commented on src/controllers/authController.js:
```diff
@@ -0,0 +1,39 @@
+ * @param {*} res - response object
+ * @return {object} - user data and status code
+ */
+  static async signupUser({
```
your code can still be refactor here is an example
static async signupUser({ body }, res) {
    try {
      const data = await AuthService.signUp(body);
      res.status(201).json({ status: 'success', data });
    } catch (error) {
     res.status(409).json({ status: 'error', error });
    }
  }

SamuelKoroh commented on src/services/authService.js:
```diff
@@ -0,0 +1,63 @@
+   * @return {JSON} - json format
+   */
+  static async signup(userDetails) {
+    const {
```
This method can also be simplified 

static async signUp(userDetails) {
    const { email, password } = userDetails;
    const checkEmail = await User.findOne({ where: { email } });

    if (checkEmail) throw new Error(`Email '${email}' already exists`);

    userDetails.password = await Helper.encryptor(password);

    const { dataValues: user } = await User.create(userDetails);
    const payload = Helper.pickFields(user, ['id', 'is_admin']);
    const token = await Helper.genToken(payload);

    return { token, ...Helper.omitFields(user, ['password']) };
  }

SamuelKoroh commented on src/services/authService.js:
```diff
@@ -0,0 +1,63 @@
+
+    if (!isPasswordValid) throw new Error('Please provide valid login credentials');
+
+    const payload = { id: user.id, is_admin: user.is_admin };
```
You can alse refactor line 27 to user the pickFields helper function
example
const payload = Helper.pickFields(user, ['id', 'is_admin']);

SamuelKoroh commented on src/test/auth.spec.js:
```diff
@@ -0,0 +1,115 @@
+  first_name: 'Samuel',
+  last_name: 'koroh',
+  email: 'user1@gmail.com',
+  password: '123456'
```
Please update the password to `password: 'Ice5m5am0a843r03'` to enable test to pass

SamuelKoroh commented on src/migrations/create-user.js:
```diff
@@ -32,7 +41,7 @@ module.exports = {
       allowNull: false,
       type: Sequelize.DATE
-    }
+    },
```
Please rename this file back to 20190815191222-create-user.js for migration to work properly.

hameedhub committed:
implementation feeback 4

Annmary12 committed:
- refactor the test script
- remove migration script from travis.yml file
- make variable names descriptive

Annmary12 committed:
add migration script in travis.yml file

Annmary12 committed:
add db migrate script

hameedhub committed:
travis error fix

hameedhub committed:
travis fix

hameedhub committed:
travis fix 2

hameedhub committed:
travis fix 3

hameedhub committed:
travis add new db

hameedhub committed:
travis db name fix

hameedhub committed:
travis undo migration

hameedhub committed:
travis migration

hameedhub committed:
travis undo create-user migration

hameedhub committed:
travis check migration status

hameedhub committed:
undo all migration

hameedhub committed:
fix migration file

hameedhub committed:
travis users relation error fix
